### DISCUSSION TITLE
legacy decorators + preset-env issue

### DISCUSSION THREAD

**Discussion by Windvis** [babel/babel#15912] [Q&A] [answered]
The [`babel-plugin-proposal-decorators` docs](https://babeljs.io/docs/babel-plugin-proposal-decorators#legacy) mention that you don't need the `@babel/plugin-transform-class-properties` and `@babel/plugin-private-methods` plugins if you are using `@babel/preset-env`.

> If you are already using @babel/preset-env, you can safely remove @babel/plugin-transform-class-properties and @babel/plugin-private-methods:

I'm trying to remove those plugins from the `ember-cli-babel` package since it also uses preset-env, but I'm seeing the following runtime exception:

> Decorating class property failed. Please ensure that transform-class-properties is enabled and runs after the decorators transform.

Which seems to get added here: https://github.com/babel/babel/blob/5ebab544af2f1c6fc6abdaae6f4e5426975c9a16/packages/babel-plugin-proposal-decorators/src/transformer-legacy.ts#L221-L224

It does work if I configure the targets to a version that requires the class fields to be transpiled.

Is the information in the documentation incorrect or am I doing something wrong? I tried reproducing it in the repl, but I'm not sure if the legacy decorators transform can be enabled there.

**Version information:**
- `@babel/core`: 7.22.11
- `@babel/preset-env`: 7.22.10
- `@babel/plugin-proposal-decorators`: 7.22.10

Any insights are appreciated, thanks!

**Comment by JLHwung** [comment, marked as answer]
The legacy decorators transform relies on the class properties / private methods transform and we have no plan to bring any feature update to that legacy mode. From now on, you have two choices:
1) migrate to the stage-3 decorator `version: "2023-05"`, which is designed to work without class elements transform. And both Babel and TS 5.0 support it well, unlike legacy decorator where they have different behaviour.
2) manually include the class elements transform with the legacy decorators

However, we should update the docs to note that you can safely remove those plugins only if you are using the stage 3 decorator transforms.

**Reply by Windvis** [reply]
Thanks for the information, that makes sense 👍.

**Reply by Windvis** [reply]
> However, we should update the docs to note that you can safely remove those plugins only if you are using the stage 3 decorator transforms.

It's especially confusing since that information is under the "legacy" option as well 😄.

**Reply by JLHwung** [reply]
Yes I agree, I am drafting a docs PR to address that.

**Reply by JLHwung** [reply]
Docs PR: https://github.com/babel/website/pull/2818, feel free to leave comments there if you have any further questions.
